gitbook/audio-streams.md

```diff
@@ -2,143 +2,129 @@
+
+```ruby
+# custom audio stream that plays a loaded buffer
+class MyStream < SF::SoundStream
+    def initialize
+        super
+        @samples = []
+        @current_sample = nil
+    end
+
+    def load(buffer)
+        # extract the audio samples from the sound buffer to our own container
+        @samples.assign(buffer.samples, buffer.samples + buffer.sample_count)
+
+        # reset the current playing position
+        @current_sample = 0
+
+        # initialize the base class
+        initialize(buffer.channel_count, buffer.sample_rate)
+    end
+
+    private def on_get_data(data)
+        # number of samples to stream every time the function is called;
+        # in a more robust implementation, it should be a fixed
+        # amount of time rather than an arbitrary number of samples
+        samples_to_stream = 50000
+
+        # set the pointer to the next audio samples to be played
+        data.samples = @samples[@current_sample]
+
+        # have we reached the end of the sound?
+        if (@current_sample + samples_to_stream <= m_samples.size())
+            # end not reached: stream the samples and continue
+            data.sample_count = samples_to_stream;
+            @current_sample += samples_to_stream;
+            true
         else
-        {
-            // end of stream reached: stream the remaining samples and stop playback
-            data.sampleCount = m_samples.size() - m_currentSample;
-            m_currentSample = m_samples.size();
-            return false;
-        }
-    }
-
-    virtual void onSeek(sf::Time timeOffset)
-    {
-        // compute the corresponding sample index according to the sample rate and channel count
-        m_currentSample = static_cast<std::size_t>(timeOffset.asSeconds() * getSampleRate() * getChannelCount());
-    }
-
-    std::vector<sf::Int16> m_samples;
-    std::size_t m_currentSample;
-};
-
-int main()
-{
-    // load an audio buffer from a sound file
-    sf::SoundBuffer buffer;
-    buffer.loadFromFile("sound.wav");
-
-    // initialize and play our custom stream
-    MyStream stream;
-    stream.load(buffer);
-    stream.play();
-
-    // let it play until it is finished
-    while (stream.getStatus() == MyStream::Playing)
-        sf::sleep(sf::seconds(0.1f));
-
-    return 0;
-}
+            # end of stream reached: stream the remaining samples and stop playback
+            data.sample_count = @samples.size - @current_sample
+            @current_sample = @samples.size
+            false
+        end
+    end
+
+    private def on_seek(time_offset)
+        # compute the corresponding sample index according to the sample rate and channel count
+        @current_sample = (time_offset.as_seconds * sample_rate * channel_count)
+    end
+end
+
+# load an audio buffer from a sound file
+buffer = SF::SoundBuffer.from_file("sound.wav")
+
+# initialize and play our custom stream
+stream = MyStream.load(buffer)
```

Comment:
Same story. `SoundStream` is an abstract class, which is really weird in CSFML and got no wrapper. Skip.